### ISSUE TITLE
DataProvider cache invalidation on locale and data changes

### ISSUE THREAD

**Issue by zbraniecki** [unicode-org/icu4x#332]
At the moment `DataProvider` is a two step state machine which is initialized and loads data. With the current bindings to APIs such as `DateTime` or `PluralRules` it allows for construction of them with a certain dataset loaded from `DataProvider`.

When that happens, the model is set and there's no way to update it in response to certain events during apps life cycle such as:

1. App locale change
2. New data

I think it would be good to discuss early on whether support for those two is in scope and if so, how will we thread through DataProvider.

For example, app locale change may require invalidation of all Formatters, and the user may do it themselves with some form of wrappers that would work like this:

```rust
struct AppCardinalPluralRules {
    get_locales: fn() -> Vec<LanguageIdentifier>,
    pr: PluralRules
}

impl AppCardinalPluralRules {
    pub fn new(get_locales: fn() -> Vec<LanguageIdentifier>) -> Self {
        let current_app_locales = App:get_locales();
        let provider = App::get_provider();
        let locale = negotiate_locales(PluralRules::available_locales(), current_app_locales);
        Self {
            get_locales,
            pr: PluralRules::try_new(locale, &provider, PluralRuleType::Cadinal)
        }
    }

    pub fn select(&self, operands: PluralOperands) -> PluralCategory {
        self.pr.select(operands)
    }

    pub fn on_locales_change(&mut self) {
        let current_app_locales = App:get_locales();
        let provider = App::get_provider();
        let locale = negotiate_locales(PluralRules::available_locales(), current_app_locales);
        self.pr = PluralRules::try_new(locale, &provider, PluralRuleType::Cardinal);
    }
}
```

This allows the app user to call `appCardinalPluralRules.select(5)` and expect it to localize using the current app locales and the result will be different after locales change.

This model is also a bit explicit asking users to wrap every instance they want to use in such wrapper (granted, the wrapper can be generalized), or never cache an instance of a formatter and recreate it on every call just because locales may sometimes change.

If we don't support it, we'll end up with "You changed your apps locales. Please restart the app to take an effect." which I think is a dirty hack.

It also doesn't resolve `new data` scenario. While we can argue that since `locale` is a manadatory constructor argument to pluralrules, it should be regenerated when the argument is different, the change to data doesn't affect the arguments passed to the constructor.

We could handle this by having `DataProvider` have some optional event listener that reacts to data changes, or have a separate crate that monitors and have methods on DataProvider to trigger an update, or have this crate regenerate DataProvider when that happens.

Then, this API will have to call all instances of such wrappers as above to regenerate them in response to new data and new data provider instance.

This solution would be modular and provide good separation of concerns, but make it more complex to handle those scenarios and require some planning of such wrappers and event triggers.

Alternatively, we could try to bake such event triggers into our existing crates effectively ending up with some form of:

```rust
// LocaleService created a DateTimeFormat using current apps locales and
// data provider.
let dtf = LocaleService::DateTimeFormat(&options).unwrap();

// Date is formatted using the current app's locale and data
dtf.format(date);

// User changed their locale fallback chain to `es, en`
LocaleService::set_requested_locales(&[langid!("es"), langid!("en")]);

// This time the date will be formatted to Spanish
dtf.format(date);

// At this point event listener reacts to `new_cldr_data` event and refreshes DataProvider and DateTimeFormats

// This time the date will be formatted to Spanish but with new cldr data
dtd.format(date);
```

I believe the users would like the ergonomics of the last example.

How do we want to design for such scenario? We need to decide what do we want the cache invalidation in those two to look like and enable API calls for that.

**Comment by sffc** [comment]
I see this as closely tied to #150. If we adopt a factory model for creating new formatters, there can be a factory with the ability to perform locale updates like this.

Caveats:

1. The core model (create formatter with locale and options) should always be available in order to have minimal dependencies
2. In all likelihood, an app will be built with a reactive UI framework (Angular, Vue, React Native) that will handle most of the events; our job is only to make it easy to plug into such frameworks
3) How does this feature interact with FFI?

**Comment by sffc** [comment]
2021-02-12: @sffc and @nciric to post comments here.  @zbraniecki to synthesize the conclusions into a markdown file in docs/papers documenting the decision on how to support this use case.

**Comment by sffc** [comment]
My comments:

Reactivity should be handled by a layer sitting on top of ICU4X (e.g., React, Angular, Vue).  When new data is available, all components that are locale-dependent should update themselves.

Caching in ICU4X should be an explicit component of the data provider pipeline.  I illustrate how a cache fits in here, in [data_pipeline.md](https://github.com/unicode-org/icu4x/blob/master/docs/design/data_pipeline.md).  Whatever event causes new locale data to become available should send a signal to the cache to invalidate its data.

Additionally, the *only* caching in ICU4X should be in the data provider pipeline.  One of the biggest design flaws in ICU is the fact that formatter construction is slow.  Having worked for nearly 5 years on ICU, I can say with some authority that ICU object construction is slow for two main reasons: (1) locale data lookup, including language negotiation + fallbacks, and (2) massaging of that locale data into a useful runtime form.  ICU4X is solving (1) by splitting the locale data provider into its own explicit component that caches data *after* language negotation and fallbacking is completed.  I am pushing strongly for solving (2) by having an ICU4X data model that requires as little processing of data at runtime as possible.  For example, ideally, the parsing of plural rules and date time patterns should be done in the CLDR-to-ICU4X transformer, not in the DateTimeFormat constructor.  We are not at this point yet, but I expect to be there when we reach v1.

In other words, in my mind, it is a design goal of ICU4X that clients *should not need to* create any other caches of ICU4X formatter objects.  And if they don't create such caches, then updating locale data throughout the stack is as simple as triggering a reactive update on the app framework layer.

**Comment by sffc** [comment]
Also CC @mihnita
